File: airbyte-integrations/bases/base-java/src/test/java/io/airbyte/integrations/util/ConnectorExceptionUtilTest.java
/*
 * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.util;

import static io.airbyte.integrations.util.ConnectorExceptionUtil.COMMON_EXCEPTION_MESSAGE_TEMPLATE;
import static io.airbyte.integrations.util.ConnectorExceptionUtil.RECOVERY_CONNECTION_ERROR_MESSAGE;
import static org.junit.jupiter.api.Assertions.*;

import io.airbyte.commons.exceptions.ConfigErrorException;
import io.airbyte.commons.exceptions.ConnectionErrorException;
import java.sql.SQLException;
import java.sql.SQLSyntaxErrorException;
import org.junit.jupiter.api.Test;

class ConnectorExceptionUtilTest {

  public static final String CONFIG_EXCEPTION_MESSAGE = "test message";
  public static final String RECOVERY_EXCEPTION_MESSAGE = "FATAL: terminating connection due to conflict with recovery";
  public static final String COMMON_EXCEPTION_MESSAGE = "something happens with connection";
  public static final String CONNECTION_ERROR_MESSAGE_TEMPLATE = "State code: %s; Error code: %s; Message: %s";
  public static final String UNKNOWN_COLUMN_SQL_EXCEPTION_MESSAGE = "Unknown column 'table.column' in 'field list'";

  @Test()
  void isConfigErrorForConfigException() {
    ConfigErrorException configErrorException = new ConfigErrorException(CONFIG_EXCEPTION_MESSAGE);
    assertTrue(ConnectorExceptionUtil.isConfigError(configErrorException));

  }

  @Test
  void isConfigErrorForConnectionException() {
    ConnectionErrorException connectionErrorException = new ConnectionErrorException(CONFIG_EXCEPTION_MESSAGE);
    assertTrue(ConnectorExceptionUtil.isConfigError(connectionErrorException));
  }

  @Test
  void isConfigErrorForRecoveryPSQLException() {
    SQLException recoveryPSQLException = new SQLException(RECOVERY_EXCEPTION_MESSAGE);
    assertTrue(ConnectorExceptionUtil.isConfigError(recoveryPSQLException));
  }

  @Test
  void isConfigErrorForUnknownColumnSQLSyntaxErrorException() {
    SQLSyntaxErrorException unknownColumnSQLSyntaxErrorException = new SQLSyntaxErrorException(UNKNOWN_COLUMN_SQL_EXCEPTION_MESSAGE);
    assertTrue(ConnectorExceptionUtil.isConfigError(unknownColumnSQLSyntaxErrorException));
  }

  @Test
  void isConfigErrorForCommonSQLException() {
    SQLException recoveryPSQLException = new SQLException(COMMON_EXCEPTION_MESSAGE);
    assertFalse(ConnectorExceptionUtil.isConfigError(recoveryPSQLException));
  }

  @Test
  void isConfigErrorForCommonException() {
    assertFalse(ConnectorExceptionUtil.isConfigError(new Exception()));
  }

  @Test
  void getDisplayMessageForConfigException() {
    ConfigErrorException configErrorException = new ConfigErrorException(CONFIG_EXCEPTION_MESSAGE);
    String actualDisplayMessage = ConnectorExceptionUtil.getDisplayMessage(configErrorException);
    assertEquals(CONFIG_EXCEPTION_MESSAGE, actualDisplayMessage);
  }

  @Test
  void getDisplayMessageForConnectionError() {
    String testCode = "test code";
    int errorCode = -1;
    ConnectionErrorException connectionErrorException = new ConnectionErrorException(testCode, errorCode, CONFIG_EXCEPTION_MESSAGE, new Exception());
    String actualDisplayMessage = ConnectorExceptionUtil.getDisplayMessage(connectionErrorException);
    assertEquals(String.format(CONNECTION_ERROR_MESSAGE_TEMPLATE, testCode, errorCode, CONFIG_EXCEPTION_MESSAGE), actualDisplayMessage);
  }

  @Test
  void getDisplayMessageForRecoveryException() {
    SQLException recoveryException = new SQLException(RECOVERY_EXCEPTION_MESSAGE);
    String actualDisplayMessage = ConnectorExceptionUtil.getDisplayMessage(recoveryException);
    assertEquals(RECOVERY_CONNECTION_ERROR_MESSAGE, actualDisplayMessage);
  }

  @Test
  void getDisplayMessageForUnknownSQLErrorException() {
    SQLSyntaxErrorException unknownColumnSQLSyntaxErrorException = new SQLSyntaxErrorException(UNKNOWN_COLUMN_SQL_EXCEPTION_MESSAGE);
    String actualDisplayMessage = ConnectorExceptionUtil.getDisplayMessage(unknownColumnSQLSyntaxErrorException);
    assertEquals(UNKNOWN_COLUMN_SQL_EXCEPTION_MESSAGE, actualDisplayMessage);
  }

  @Test
  void getDisplayMessageForCommonException() {
    Exception exception = new SQLException(COMMON_EXCEPTION_MESSAGE);
    String actualDisplayMessage = ConnectorExceptionUtil.getDisplayMessage(exception);
    assertEquals(String.format(COMMON_EXCEPTION_MESSAGE_TEMPLATE, COMMON_EXCEPTION_MESSAGE), actualDisplayMessage);
  }

  @Test
  void getRootConfigErrorFromConfigException() {
    ConfigErrorException configErrorException = new ConfigErrorException(CONFIG_EXCEPTION_MESSAGE);
    Exception exception = new Exception(COMMON_EXCEPTION_MESSAGE, configErrorException);

    Throwable actualRootConfigError = ConnectorExceptionUtil.getRootConfigError(exception);
    assertEquals(configErrorException, actualRootConfigError);
  }

  @Test
  void getRootConfigErrorFromRecoverySQLException() {
    SQLException recoveryException = new SQLException(RECOVERY_EXCEPTION_MESSAGE);
    RuntimeException runtimeException = new RuntimeException(COMMON_EXCEPTION_MESSAGE, recoveryException);
    Exception exception = new Exception(runtimeException);

    Throwable actualRootConfigError = ConnectorExceptionUtil.getRootConfigError(exception);
    assertEquals(recoveryException, actualRootConfigError);
  }

  @Test
  void getRootConfigErrorFromUnknownSQLErrorException() {
    SQLException unknownSQLErrorException = new SQLSyntaxErrorException(UNKNOWN_COLUMN_SQL_EXCEPTION_MESSAGE);
    RuntimeException runtimeException = new RuntimeException(COMMON_EXCEPTION_MESSAGE, unknownSQLErrorException);
    Exception exception = new Exception(runtimeException);

    Throwable actualRootConfigError = ConnectorExceptionUtil.getRootConfigError(exception);
    assertEquals(unknownSQLErrorException, actualRootConfigError);
  }

  @Test
  void getRootConfigErrorFromNonConfigException() {
    SQLException configErrorException = new SQLException(CONFIG_EXCEPTION_MESSAGE);
    Exception exception = new Exception(COMMON_EXCEPTION_MESSAGE, configErrorException);

    Throwable actualRootConfigError = ConnectorExceptionUtil.getRootConfigError(exception);
    assertEquals(exception, actualRootConfigError);
  }

}

File: airbyte-cdk/java/airbyte-cdk/core/src/test/kotlin/io/airbyte/cdk/integrations/util/ConnectorExceptionUtilTest.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.util

import io.airbyte.commons.exceptions.ConfigErrorException
import io.airbyte.commons.exceptions.ConnectionErrorException
import java.sql.SQLException
import java.sql.SQLSyntaxErrorException
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

internal class ConnectorExceptionUtilTest {
    @get:Test
    val isConfigErrorForConfigException: Unit
        get() {
            val configErrorException = ConfigErrorException(CONFIG_EXCEPTION_MESSAGE)
            Assertions.assertTrue(ConnectorExceptionUtil.isConfigError(configErrorException))
        }

    @get:Test
    val isConfigErrorForConnectionException: Unit
        get() {
            val connectionErrorException = ConnectionErrorException(CONFIG_EXCEPTION_MESSAGE)
            Assertions.assertTrue(ConnectorExceptionUtil.isConfigError(connectionErrorException))
        }

    @get:Test
    val isConfigErrorForRecoveryPSQLException: Unit
        get() {
            val recoveryPSQLException = SQLException(RECOVERY_EXCEPTION_MESSAGE)
            Assertions.assertTrue(ConnectorExceptionUtil.isConfigError(recoveryPSQLException))
        }

    @get:Test
    val isConfigErrorForUnknownColumnSQLSyntaxErrorException: Unit
        get() {
            val unknownColumnSQLSyntaxErrorException =
                SQLSyntaxErrorException(UNKNOWN_COLUMN_SQL_EXCEPTION_MESSAGE)
            Assertions.assertTrue(
                ConnectorExceptionUtil.isConfigError(unknownColumnSQLSyntaxErrorException)
            )
        }

    @get:Test
    val isConfigErrorForCommonSQLException: Unit
        get() {
            val recoveryPSQLException = SQLException(COMMON_EXCEPTION_MESSAGE)
            Assertions.assertFalse(ConnectorExceptionUtil.isConfigError(recoveryPSQLException))
        }

    @get:Test
    val isConfigErrorForCommonException: Unit
        get() {
            Assertions.assertFalse(ConnectorExceptionUtil.isConfigError(Exception()))
        }

    @get:Test
    val displayMessageForConfigException: Unit
        get() {
            val configErrorException = ConfigErrorException(CONFIG_EXCEPTION_MESSAGE)
            val actualDisplayMessage =
                ConnectorExceptionUtil.getDisplayMessage(configErrorException)
            Assertions.assertEquals(CONFIG_EXCEPTION_MESSAGE, actualDisplayMessage)
        }

    @get:Test
    val displayMessageForConnectionError: Unit
        get() {
            val testCode = "test code"
            val errorCode = -1
            val connectionErrorException =
                ConnectionErrorException(testCode, errorCode, CONFIG_EXCEPTION_MESSAGE, Exception())
            val actualDisplayMessage =
                ConnectorExceptionUtil.getDisplayMessage(connectionErrorException)
            Assertions.assertEquals(
                String.format(
                    CONNECTION_ERROR_MESSAGE_TEMPLATE,
                    testCode,
                    errorCode,
                    CONFIG_EXCEPTION_MESSAGE
                ),
                actualDisplayMessage
            )
        }

    @get:Test
    val displayMessageForRecoveryException: Unit
        get() {
            val recoveryException = SQLException(RECOVERY_EXCEPTION_MESSAGE)
            val actualDisplayMessage = ConnectorExceptionUtil.getDisplayMessage(recoveryException)
            Assertions.assertEquals(
                ConnectorExceptionUtil.RECOVERY_CONNECTION_ERROR_MESSAGE,
                actualDisplayMessage
            )
        }

    @get:Test
    val displayMessageForUnknownSQLErrorException: Unit
        get() {
            val unknownColumnSQLSyntaxErrorException =
                SQLSyntaxErrorException(UNKNOWN_COLUMN_SQL_EXCEPTION_MESSAGE)
            val actualDisplayMessage =
                ConnectorExceptionUtil.getDisplayMessage(unknownColumnSQLSyntaxErrorException)
            Assertions.assertEquals(UNKNOWN_COLUMN_SQL_EXCEPTION_MESSAGE, actualDisplayMessage)
        }

    @get:Test
    val displayMessageForCommonException: Unit
        get() {
            val exception: Exception = SQLException(COMMON_EXCEPTION_MESSAGE)
            val actualDisplayMessage = ConnectorExceptionUtil.getDisplayMessage(exception)
            Assertions.assertEquals(
                String.format(
                    ConnectorExceptionUtil.COMMON_EXCEPTION_MESSAGE_TEMPLATE,
                    COMMON_EXCEPTION_MESSAGE
                ),
                actualDisplayMessage
            )
        }

    @get:Test
    val rootConfigErrorFromConfigException: Unit
        get() {
            val configErrorException = ConfigErrorException(CONFIG_EXCEPTION_MESSAGE)
            val exception = Exception(COMMON_EXCEPTION_MESSAGE, configErrorException)

            val actualRootConfigError = ConnectorExceptionUtil.getRootConfigError(exception)
            Assertions.assertEquals(configErrorException, actualRootConfigError)
        }

    @get:Test
    val rootConfigErrorFromRecoverySQLException: Unit
        get() {
            val recoveryException = SQLException(RECOVERY_EXCEPTION_MESSAGE)
            val runtimeException = RuntimeException(COMMON_EXCEPTION_MESSAGE, recoveryException)
            val exception = Exception(runtimeException)

            val actualRootConfigError = ConnectorExceptionUtil.getRootConfigError(exception)
            Assertions.assertEquals(recoveryException, actualRootConfigError)
        }

    @get:Test
    val rootConfigErrorFromUnknownSQLErrorException: Unit
        get() {
            val unknownSQLErrorException: SQLException =
                SQLSyntaxErrorException(UNKNOWN_COLUMN_SQL_EXCEPTION_MESSAGE)
            val runtimeException =
                RuntimeException(COMMON_EXCEPTION_MESSAGE, unknownSQLErrorException)
            val exception = Exception(runtimeException)

            val actualRootConfigError = ConnectorExceptionUtil.getRootConfigError(exception)
            Assertions.assertEquals(unknownSQLErrorException, actualRootConfigError)
        }

    @get:Test
    val rootConfigErrorFromNonConfigException: Unit
        get() {
            val configErrorException = SQLException(CONFIG_EXCEPTION_MESSAGE)
            val exception = Exception(COMMON_EXCEPTION_MESSAGE, configErrorException)

            val actualRootConfigError = ConnectorExceptionUtil.getRootConfigError(exception)
            Assertions.assertEquals(exception, actualRootConfigError)
        }

    companion object {
        const val CONFIG_EXCEPTION_MESSAGE: String = "test message"
        const val RECOVERY_EXCEPTION_MESSAGE: String =
            "FATAL: terminating connection due to conflict with recovery"
        const val COMMON_EXCEPTION_MESSAGE: String = "something happens with connection"
        const val CONNECTION_ERROR_MESSAGE_TEMPLATE: String =
            "State code: %s; Error code: %s; Message: %s"
        const val UNKNOWN_COLUMN_SQL_EXCEPTION_MESSAGE: String =
            "Unknown column 'table.column' in 'field list'"
    }
}
